Proptest-derive book (#104)
AltSysrq/proptest: book/src/proptest-derive/modifiers.md

# Modifier Reference

All modifiers interpreted by `#[derive(Arbitrary)]` are of the form
`#[proptest(..)]`, where the content between the parentheses follows the normal
Rust attribute syntax.

Each modifier within the parentheses is independent, in that putting two
modifiers in the same attribute is equivalent to having two `#[proptest(..)]`
attributes with one modifier each.

For brevity, modifiers are sometimes referenced by name alone; e.g., "the
`weight` modifier" refers to `#[proptest(weight = nn)]` and not some

> freestanding `#[weight]` attribute.

Reviewer: Hmm... I think `weight = W` works, but a reader might now have read this bit and is jumping around and so `#[weight]` could be confusing...

## `filter`

Form: `#[proptest(filter = F)]` or `#[proptest(filter(F))]` where `F` is either

> a bare identifier or a Rust expression in a string. In either case, the

Reviewer: Clarify that "bare identifier" refers to a function (or const/static also works...)?

> parameter must evaluate to something which is `Fn (&T) -> bool`, where `T` is

Reviewer: ```suggestion
parameter must evaluate to something which is `Fn(&T) -> bool`, where `T` is
```

the type of the item being filtered.

Usable on: structs, enums, enum variants, fields

The `filter` modifier allows filtering values generated for a field via
rejection sampling. Since rejection sampling is inefficient and interferes with
shrinking, it should only be used for conditions that are very rare or are
unfeasible to express otherwise. In many cases, [`strategy`](#strategy) can be
used to more directly express the desired behaviour without rejection sampling.
See the documentation for [`prop_filter`] for more details.

The argument to the modifier must be a valid argument for the second parameter
of [`prop_filter`].

Example:

```rust
#[derive(Debug, Arbitrary)]
#[proptest(filter = "|segment| segment.start != segment.end")]
struct NonEmptySegment {
    start: i32,
    end: i32,
}
```

As mentioned above, filtering should be avoided when it is reasonably possible
to express a non-filtering strategy that achieves the same effect. For example:

```rust
#[derive(Debug, Arbitrary)]

> struct BadExapmle {

Reviewer: ```suggestion
struct BadExample {
```

    // Don't do this! Your tests will run more slowly and shrinking won't work
    // properly.

>     #[proptest(filter = "|x| x % 2 == 0")]

Reviewer: Maybe extract this out to an `is_even` function to demonstrate how that works?

Author: I added a separate example to the main usage example.

    even_number: u32,
}

#[derive(Debug, Arbitrary)]
struct GoodExample {
    // Directly generate even numbers only by transforming the set of all
    // `u32`s and then mapping it to the set of even `u32`s.
    #[proptest(strategy = "any::<u32>().prop_map(|x| x / 2 * 2)")]
    even_number: u32,
}
```

[`prop_filter`]: https://altsysrq.github.io/rustdoc/proptest/latest/proptest/strategy/trait.Strategy.html#method.prop_filter

## `no_bound`

Form: `#[proptest(no_bound)]`

> Usable on: generic type declarations

Reviewer: ```suggestion
Usable on: generic type definitions and type parameters
```

Author: I had not noticed that was supported. Added it to the docs.

Normally, when `#[derive(Arbitrary)]` is applied to an item with generic type

> arguments, every type argument is required to `impl Arbitrary`. For example,

Reviewer: ```suggestion
parameters, every type parameter is required to `impl Arbitrary`. For example,
```

given a declaration like the following:

```rust
#[derive(Arbitrary)]
struct MyStruct<T> { /* ... */ }
```

Something like this will be generated:

```rust
impl<T> Arbitrary for MyStruct<T> where T: Arbitrary { /* ... */ }
```

Placing `#[proptest(no_bound)]` on a generic type argument suppresses this. For
example, the following removes the extra `where T: Arbitrary`:

```rust
#[derive(Arbitrary)]
struct MyStruct<#[proptest(no_bound)] T> { /* ... */ }

> ```

Reviewer: I'd also include an example:

```rust
#[derive(Arbitrary)]
#[proptest(no_bound)]
struct MyStruct<T, U> { /* ... */ }
```

since that short-hand also exists.

> In order for `no_bound` to be useful, one of two things needs to be true:

Reviewer: ```suggestion
For `no_bound` to be useful, one of two things needs to be true:
```

- The nature of the item is such that no value with the generic type actually
  comes into existence. This can happen if the only references are things such
  as `PhantomData<T>` or if the item is an enum and all variants referencing
  the type are marked `#[proptest(skip)]`.

- Other modifiers (such as `value`) are added to things referencing the generic
  type that remove proptest's responsibility for generating them.

Example of the above:

```rust
#[derive(Debug, Arbitrary)]
enum MyEnum<#[proptest(no_bound)] T> {
    // OK -- no T needs to be constructed

>     Phantom(PhantomData<T>),

Reviewer: The macro recognizes `PhantomData<T>` ("magic", see `if util::is_phantom_data(path) { return; }`) and will not add a bound for that. See tests in `tests/phantom.rs`.

Author: Hm, I'm not sure how I feel about that… even the built in derives don't do that.

    // OK -- we're manually taking responsibility for generating the value
    #[proptest(value = "MyEnum::WithValue(None)")]

>     WithValue(Option<T>),

Reviewer: A consequence of taking responsibility is that `T: Arbitrary` goes away as well.

    // OK -- we tell proptest not to generate this case
    #[proptest(skip)]

>     Skipped(T),

Reviewer: A consequence of skipping is that `T: Arbitrary` won't be added also.

    // ERROR -- proptest would need to generate a T here, but cannot because
    // we lack the `T: Arbitrary` bound due to having used `no_bound`.

>     Error(T),

Reviewer: But this is correct; In general, when you make the macro use `any` or `any_with` then `T: Arbitrary` will be added... You can see this by looking for `.mark_uses(` and some of the logic inside `UseTracker::add_bounds`.

This also applies when you use some associated type... If we we write (`tests/assoc.rs`):
```rust
#[derive(Debug, Arbitrary)]
struct T4<Tyvar: FuncB> {
    field: Tyvar::OutB,
}
```
then we require `<Tyvar as FuncB>::OutB: Arbitrary` but not `Tyvar: Arbitrary`.

The following situation would require `no_bound` tho:

```rust
#[derive(Debug, Arbitrary)]
struct Hide<T>(PhantomData<T>);

#[derive(Debug, Arbitrary)]
struct Stuff<T>(Hide<T>);
// We have no idea that `Arbitrary for Hide<T>` doesn't require `T: Arbitrary`
// so we use `no_bound` to get rid of the extra bound.
```

I should probably add a new round of tests to ensure the semantics are as expected and more robust in the face of changes. Some documentation improvements in the code itself are also probably necessary on my part.

}
```

## `no_params`

> Form: `#[no_params]`

Reviewer: ```suggestion
Form: `#[proptest(no_params)]`
```

I'd write it out fully; saves some confusion in some cases imo.

`Default::default()` as there is no automatic way to locate an appropriate
value (if there even is any) within the `params` type.

If applied to an enum variant or field, `params` specifies the parameters type
for just that item, as if its type had an `Arbitrary` implementation taking
that type. In this case, either [`value`](#value) or [`strategy`](#strategy)
_must_ be specified since the parameter type will not generally be compatible
with the normal `Arbitrary` invocation (and in cases where it is, `params`
would be useless if not used).

Any expressions (such as in the [`value`](#value) and [`strategy`](#strategy)
modifiers) underneath an item with the `params` modifier has access to a
variable named `params` which is of the type passed in
`#[proptest(params = ..)]`.

Examples:

```rust
#[derive(Debug)]
struct WidgetRange(usize, usize);

impl Default for WidgetRange {
    fn default() -> Self { Self(0, 100) }
}

#[derive(Debug, Arbitrary)]
#[proptest(params(WidgetRange))]
struct WidgetCollection {
    #[proptest(strategy = "params.0 ..= params.1")]
    desired_widget_count: usize,
    // ...
}

// ...

proptest! {
    #[test]

>     fn test_something(wc in any::<WidgetCollection>(WidgetRange(10, 20))) {

Reviewer: ```suggestion
    fn test_something(wc in any_with::<WidgetCollection>(WidgetRange(10, 20))) {
```

`string` is a regular expression. May also be invoked as
`#[proptest(regex(function_name))]`, where `function_name` is a no-argument
function that returns an `&'static str`.

Usable on: fields

This modifier specifies to generate character or byte strings for a field which
match a particular regular expression.

The `regex` modifier is equivalent to using the [`strategy`](#strategy) modifier and
enclosing the string in [`string_regex`] or [`bytes_regex`]. It can only be
applied to fields of type `String` or `Vec<u8>`.

Example:

```rust
#[derive(Debug, Arbitrary)]
struct FileContent {
    #[regex = "[a-z0-9.]+"]
    name: String,
    #[regex = "([0-9]+\n)*"]
    content: Vec<u8>,
}
```

[`string_regex`]: https://altsysrq.github.io/rustdoc/proptest/latest/proptest/string/fn.string_regex.html
[`bytes_regex`]: https://altsysrq.github.io/rustdoc/proptest/latest/proptest/string/fn.bytes_regex.html

## `skip`

Form: `#[proptest(skip)]`

Usable on: enum variants

Annotating an enum variant with `#[proptest(skip)]` prevents proptest from
generating that particular variant. This is useful when there is no sensible

> way to generate the variant.

Reviewer: ```suggestion
way to generate the variant or when you want to temporarily stop generating some variant during development.
```

`#[proptest(skip)]` as this leaves proptest with no options to generate the
enum.

## `strategy`

Form: `#[proptest(strategy = S)]` or `#[proptest(strategy = S)]`, where `S` is
either a string containing a Rust expression which evaluates to an appropriate
`Strategy`, or a bare identifier naming a function which, when called with no
arguments, returns such a `Strategy`.

Usable on: enum variants, fields

By default, enum variants are generated by recursing into their definition as
is done for struct declarations, and fields are generated by invoking
`Arbitrary` on the field type to produce a `Strategy`. The `strategy` modifier
allows to manually provide a custom strategy directly.

In the case of fields, the strategy must produce values of the same type as
that field. For enum variants, it must produce values of the enum type itself
and these values ought to be of the variant in question.

Example:

```rust
#[derive(Debug, Arbitrary)]
enum Token {
    Delimitation {
        // This field is still generated via Arbitrary
        delimiter: Delimiter,

        // But for this field we use a custom strategy
        #[proptest(strategy = "1..10")]

>         count: u32,

Reviewer: Maybe add an example of a bare identifier as an extra field?

function to produce a value for the field, instead of going through the usual
value generation machinery.

The argument to `value` is directly used as an expression for the field value
or enum variant to be generated, except that in the third form where it is a
bare identifier, it is called as a no-argument function to produce the value.

Using `value` is equivalent to using [`strategy`](#strategy) and enclosing the
value in `LazyJust`.

Example:

```rust
#[derive(Debug, Arbitrary)]
struct EventCounter {
    // We always start with the first two fields set to 0/None
    #[proptest(value = 0)]
    number_seen: u64,

    #[proptest(value = "None")]
    last_seen_time: Option<Instant>,

    // This field is generated normally
    max_events: u64,
}
```

## `weight`

Form: `#[proptest(weight = W)]` or `#[proptest(weight(W))]`, where `W` is an
expression evaluating to a `u32`. `weight` may also be abbreviated to `w`, as
in `#[proptest(w = W)]`.

Usable on: enum variants

The `weight` modifier determines how likely proptest is to generate a
particular enum variant. Weights are relative to each other; for example, a

> `weight=3` variant is 50% more likely to be generated than a `weight=2` variant

Reviewer: ```suggestion
`weight = 3` variant is 50% more likely to be generated than a `weight = 2` variant
```

> and three times as likely to be generated as a `weight=1` variant.

Reviewer: ```suggestion
and three times as likely to be generated as a `weight = 1` variant.
```
